pkgs/build-support/node/build-npm-package/default.nix

```
@@ -0,0 +1,35 @@
+{ lib, stdenv, fetchNpmDeps, callPackage, nodejs }:
+
+{ name ? "${args.pname}-${args.version}"
+, src ? null
+, srcs ? null
+, sourceRoot ? null
+, patches ? [ ]
+, npmDepsHash ? ""
+, nativeBuildInputs ? [ ]
+, buildInputs ? [ ]
+, ...
+} @ args:
+
+assert npmDepsHash != "";
+
+let
+  npmDeps = fetchNpmDeps {
+    inherit src srcs sourceRoot patches;
+    name = "${name}-npm-deps";
+    hash = npmDepsHash;
+  };
+
+  # TODO: how do we expose these like Rust? `npmHooks`?
+  inherit (callPackage ./hooks { inherit nodejs; }) npmConfigHook npmBuildHook npmInstallHook;
+in
+stdenv.mkDerivation (args // {
+  inherit npmDeps;
+
+  nativeBuildInputs = nativeBuildInputs ++ [ nodejs npmConfigHook npmBuildHook npmInstallHook ];
+  buildInputs = buildInputs ++ [ nodejs ];
```

Why is `nodejs` in both `buildInputs` and `nativeBuildInputs`? Since `nodeHost` gets wrapped by absolute path in `installPhase` we shouldn't need it in `buildInputs` I don't think.

Also prefer this formatting so diffs are easier to read if we add/change an input:
```suggestion
  nativeBuildInputs = nativeBuildInputs ++ [
    nodejs
    npmConfigHook
    npmBuildHook
    npmInstallHook
  ];

  buildInputs = buildInputs ++ [
    nodejs
  ];
```

---

> Why is `nodejs` in both `buildInputs` and `nativeBuildInputs`?

For `patchShebangs` in `fixupPhase`, just in case.

---

Interesting, i wasn't aware shebangs are ever used sith java files.